Add a diff view before saving changes in the ScopeEditor

MynahMarie opened the pull request:
In this PR I have:

- added a 4th button to the `ButtonToolbar` which toggles between "**Review**" and "**Edit**"
- created a new rendering function that shows the changes the user made if he/she wishes to review them before saving.
- created a customized version of the `react-diff` module in order to create a component that renders those changes
- created a new directory in the project's root to store the custom `react-diff` module I've created

Relates Bugzilla 1306494

eliperelman commented on custom_modules/react-diff-custom/dist/react-diff-custom.js:
```diff
@@ -0,0 +1,59 @@
+'use strict';
```
We already have a `components` directory inside `src`, and this can go there. Some things to change:

- Use ES `import` and `export` instead of CommonJS `require` and `module.exports`.
- Use `const` over `var`, which the linter will enforce once this is moved into `src`.
- `React.createClass` isn't used in this codebase, as ES classes are used. See all other components in this repo for examples.
- React isn't going to have `PropTypes` on it, as we are importing the `prop-types` package.
- Use JSX.

djmitche replied:
Another option here is to try to contribute these changes to the upstream package or, if that doesn't work or the changes are so massive that it doesn't make sense, create a new package on npm for this.

MynahMarie replied:
@eliperelman So the first thing I tried was to add it in the src folder but there was 1 linting error I just didn't manage to get rid of... I will try again with more intense re-writing and see if I can get it to work.

@djmitche yes I thought about that also... I looked at the react-diff project's repo and not much activity is happening there and there is an issue of the owner saying that he is looking for someone to maintain it. I think I will try and see if I can make it work as Eli suggested and if that doesn't work I will try to contact the owner of the module directly.

eliperelman commented on custom_modules/react-diff-custom/dist/react-diff-custom.js:
```diff
@@ -0,0 +1,59 @@
+'use strict';
+
+var React = require('react');
```
What is different about this file from the original that couldn't be fixed via different props?

MynahMarie replied:
The only things I changed from the original file was to change the styling and to add the option of using the `diffLines` method from the `diff` module. Using `diffLines` allows to pass it the `newlineIsToken: true` option which I felt was the most appropriate one for rendering the kind of diff we needed...

eliperelman commented on custom_modules/react-diff-custom/package.json:
```diff
@@ -0,0 +1,39 @@
+{
```
This file isn't needed once the component lives in `src`.

MynahMarie commented:
@eliperelman I moved the `react-diff-custom` to the components folder in `src` and deleted the `package.json` associated to it. I also refactored using `jsx` and ES6 syntax, please let me know if this is an improvement and if there is any other changes I should do.

eliperelman commented on src/components/ScopeEditor/index.jsx:
```diff
@@ -8,12 +8,14 @@ import 'codemirror/addon/mode/simple';
 import CodeEditor from '../CodeEditor';
 import './scopemode';
 import './styles.css';
+import Diff from './react-diff-custom';
```
For consistency, this file should be in PascalCase like all other components. You can just name it `Diff`.

eliperelman commented on src/components/ScopeEditor/react-diff-custom.jsx:
```diff
@@ -0,0 +1,68 @@
+// This implementation is a customization of the 'react-diff' module.
```
Can you move these comments to below the imports?

eliperelman commented on src/components/ScopeEditor/react-diff-custom.jsx:
```diff
@@ -0,0 +1,68 @@
+// This implementation is a customization of the 'react-diff' module.
+// https://github.com/cezary/react-diff
+
+import { Component, PropTypes } from 'react';
```
We don't use `PropTypes` from the react module as they have been removed/deprecated. For consistency, see how we import prop types in all other components: by importing from the `prop-types` package.

eliperelman commented on src/components/ScopeEditor/react-diff-custom.jsx:
```diff
@@ -0,0 +1,68 @@
+    type: PropTypes.oneOf(['chars', 'words', 'sentences', 'json', 'lines'])
+  };
+
+  getDefaultProps = () => ({
```
You don't need a function to get the default props, and additionally I don't see this function being used. Declare a static default props property:

```js
static defaultProps = {
  inputA: '',
  inputB: '',
  type: 'chars'
};
```

eliperelman commented on src/components/ScopeEditor/react-diff-custom.jsx:
```diff
@@ -0,0 +1,68 @@
+
+export default class Diff extends Component {
+  static propTypes = {
+    inputA: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
```
These variables names could be improved to convey the intent behind them. For example, instead of `inputA`, we could name it `original`, and `inputB` could be named `updated`.

eliperelman commented on src/components/ScopeEditor/react-diff-custom.jsx:
```diff
@@ -0,0 +1,68 @@
+  });
+
+  render() {
+    let diff = fnMap[this.props.type](this.props.inputA, this.props.inputB);
```
This can be a ternary expression which would lend itself to being a `const`-preferred declaration:

```js
const diff = this.props.propTypes === 'lines'
  ? fnMap[this.props.type](
      this.props.inputA,
      this.props.inputB[{ newlineIsToken: true }]
    )
  : fnMap[this.props.type](this.props.inputA, this.props.inputB);
```

eliperelman commented on src/components/ScopeEditor/react-diff-custom.jsx:
```diff
@@ -0,0 +1,68 @@
+      );
+    }
+
+    // Change the original styling.
```
What was the original styling?

eliperelman commented on src/components/ScopeEditor/react-diff-custom.jsx:
```diff
@@ -0,0 +1,68 @@
+      };
+
+      return (
+        <span key={index} style={spanStyle}>
```
We cannot use the `index` to drive the keys here as we are doing line sorting, which means we can end up in a situation where two lines of different content end up with the same key and cause rendering issues. You'll need to come up with a way to derive a key based on the content of the line.

eliperelman commented on src/components/ScopeEditor/react-diff-custom.jsx:
```diff
@@ -0,0 +1,68 @@
+      let color = 'grey';
+      let backgroundColor = null;
+
+      if (part.added) {
```
These CSS styles can be extracted into class names using CSS modules. See other uses of CSS files in this repo for some examples of how to do that, as opposed to using inline styles.

eliperelman commented on src/components/ScopeEditor/react-diff-custom.jsx:
```diff
@@ -0,0 +1,68 @@
+      );
+    });
+
+    return <pre className="diff-result">{result}</pre>;
```
I don't see this class name used anywhere. Does it have some purpose with the library, or is it just added here for identification?

eliperelman commented on src/components/ScopeEditor/react-diff-custom.jsx:
```diff
@@ -0,0 +1,68 @@
+  lines: jsdiff.diffLines
+};
+
+export default class Diff extends Component {
```
As mentioned elsewhere, since this component is named `Diff`, the filename should be `Diff.jsx`. Also, this doesn't seem to be specific to scopes, so we can move it to the `components` directory as a top-level item as `components/Diff/index.jsx`. That would give us the ability to use diff viewers in more places than this in the future.

eliperelman commented on src/components/ScopeEditor/index.jsx:
```diff
@@ -83,6 +86,16 @@ export default class ScopeEditor extends Component {
+  renderDiff() {
+    return (
+      <Diff
+        inputA={this.state.oldScopes.join('\n')}
```
In the original bug we discussed the need for sorting the scopes prior to rendering them or showing them in the diff viewer so the diff would be accurate. Is that not needed or helpful?

MynahMarie commented:
@eliperelman Hi, I am really sorry for not continuing with this sooner, I had a contract that needed all my attention and had not time... I just recently got back to working on it, I've addressed most of the requested changes (the comments I gave a thumbs up are the changes I've addressed, still missing a few). Is it ok if I continue to work on this or is this PR now too outdated?

eliperelman commented:
> Is it ok if I continue to work on this or is this PR now too outdated?

No, feel free to continue. 😃

helfi92 commented:
@MynahMarie Let us know when this is ready to be reviewed :)
